astropy/utils/console.py

```diff
@@ -899,73 +899,80 @@ def __init__(self, msg, color='default', file=None, step=1,
         self._silent = not isatty(file)
 
+        if self._silent:
+            self._iter = self._silent_iterator
+        else:
+            self._iter = self._iterator
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        file = self._file
+        write = file.write
+        flush = file.flush
+
+        if not self._silent:
+            write('\r')
+            color_print(self._msg, self._color, file=file, end='')
+        if exc_type is None:
+            color_print(' [Done]', 'green', file=file)
+        else:
+            color_print(' [Failed]', 'red', file=file)
+        flush()
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        self._iter()
+
     def _iterator(self):
+        """
+        Update the spin wheel.
+        """
         chars = self._chars
-        index = 0
         file = self._file
         write = file.write
-        flush = file.flush
         try_fallback = True
 
-        while True:
+        if self._cur_step % self._step == 0:
             write('\r')
             color_print(self._msg, self._color, file=file, end='')
             write(' ')
             try:
                 if try_fallback:
-                    write = _write_with_fallback(chars[index], write, file)
+                    write = _write_with_fallback(chars[self._cur_char], write, file)
                 else:
-                    write(chars[index])
+                    write(chars[self._cur_char])
             except UnicodeError:
                 # If even _write_with_fallback failed for any reason just give
                 # up on trying to use the unicode characters
                 chars = self._default_ascii_chars
-                write(chars[index])
+                write(chars[self._cur_char])
                 try_fallback = False  # No good will come of using this again
-            flush()
-            yield
-
-            for i in range(self._step):
-                yield
-
-            index = (index + 1) % len(chars)
-
-    def __enter__(self):
-        if self._silent:
-            return self._silent_iterator()
-        else:
-            return self._iterator()
-
-    def __exit__(self, exc_type, exc_value, traceback):
-        file = self._file
-        write = file.write
-        flush = file.flush
+            self._file.flush()
 
-        if not self._silent:
-            write('\r')
-            color_print(self._msg, self._color, file=file, end='')
-        if exc_type is None:
-            color_print(' [Done]', 'green', file=file)
-        else:
-            color_print(' [Failed]', 'red', file=file)
-        flush()
+            self._cur_char = (self._cur_char + 1) % len(chars)
+        self._cur_step = self._cur_step + 1
 
     def _silent_iterator(self):
         color_print(self._msg, self._color, file=self._file, end='')
         self._file.flush()
 
-        while True:
-            yield
+    def update(self):
```

Comment:
This should have a docstring.

Reply:
Although not as useful as it is in a progress bar, for symmetry it would be nice if this also took an optional `value` argument.

You could write:

```
def update(self, value=None):
    if value is None:
        value = 1
    for _ in range(value):
        next(self)
```

Then that would allow you to do away entirely with this conditional: https://github.com/astropy/astropy/pull/11015/files#diff-582a94e0a0f7293634d5bb6531b7bdf8a847ecd2a380ec691a65bac8beaff546R1048

Reply:
Perhaps its better to just write `def update(self, value=1)` ?